master/buildbot/steps/trigger.py

```diff
@@ -297,6 +306,10 @@ def run(self):
                 # the beginning
                 url = self.master.status.getURLForBuildrequest(brid)
                 yield self.addURL("{} #{}".format(sch.name, brid), url)
+                deferredRes = self.master.mq.waitUntilEvent(('buildrequests', str(brid),
```

Comment:
I think it would be more readable if this was extracted into a separate `@defer.inlineCallbacks function` which is not awaited via `yield` here (add a comment explaining this). I guess that function could be merged with `addResults()`.

Reply:
I still struggle to get twisted code right so if I'm changing this I'll need to look at it more carefully to double check I understand what you mean. I've made the other tweaks though.

Reply:
You can have a look at basic asyncio tutorial to get a grasp about it.

for example: https://stackabuse.com/python-async-await-tutorial/

twisted has the same concepts but 10 years before, so the API is not as nice.

but basically you can say

```python
@defer.inlineCallbacks
def method(...)
```

is equivalent to

```python
async def method(...)
```
and

```python
     yield self.returningDeferredFunction()
```
is equivalent to
```python
    await self.returningFutureOrCoroutineFunction()
```

There are some subtle differences, but most of the time, this is not relevent.

The key learning is: don't use callbacks, there are barely any case where it is useful.
There is only a slight performance gain, and it can be easier to reason on for hardcore network protocol stuff. But for 99% of buildbot code, we want to switch to inlineCallbacks (the coroutines of twisted)

Reply:
Thanks! I think I understand the basics but what I was struggling with here was how to have a callback function which didn't block the rest of the code (so no yield) yet used the inline callbacks. I'm sure if I stare at it further I can work it out, I just ran out of time yesterday...

Reply:
ah. Indeed, this is the 1% case where you want to create a callback.

As it is not obvious, I guess we'll just need a simple comment.

I think what @p12tic was suggesting is to just use fire and forget

```
    @defer.inlineCallbacks
    def _wait_for_results(self, brid):
        br, brid = yield self.master.mq.waitUntilEvent(('buildrequests', str(brid), 'complete', is_buildrequest_complete(brid)
        builds = yield self.master.db.builds.getBuilds(buildrequestid=brid)
        for build in builds:
            self._result_list.append(build["results"])
        self.updateSummary()
```
and then in run() method, you just call this method and ignore the returning deferred.

Note, that you will need to implement is_buildrequest_complete, by just calling the data api, in order to avoid race condition between actual completion and event registering

Reply:
The is_buildrequest_complete(brid) here threw me for a while, it needs to be is_buildrequest_complete, brid :)

Reply:
just edited my previous comment to add the needed lambda:

Reply:
Thanks, I got there as well and have updated the pull request. Might be a style issue left on where the function is located but hopefully getting there!